### PR TITLE
EMQX Agent Integration

### PR DESCRIPTION
**Pull request by zhongwencool**
### What does this PR do?

Integrate [MQTT broker EMQX](https://www.emqx.com/en/blog/the-ultimate-guide-to-mqtt-broker-comparison)  

### Motivation

What inspired you to submit this pull request?

### Review checklist

- [ ] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [ ] Feature or bugfix has tests
- [ ] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:5]: Can you please provide a bit more context on:
- What EMQX is/what it does
- The type of data this integration sends into Datadog/why it's valuable for the user

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:16]:
```suggestion
Then, manually install the EMQX check. [Instructions vary depending on the environment][10].
```
Is this the correct link?

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:18]:
```suggestion
Run `datadog-agent integration install -t datadog-emqx==1.0.0`.
```

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:22]: Typo?
```suggestion
1. Edit the `emqx/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your EMQX performance data.
```

File: emqx/assets/dashboards/overview.json
Comment (anawishnoff) [inline emqx/assets/dashboards/overview.json:1]: A few requests on this dashboard:
1. Please add your logo to the dashboard in the top left corner
2. Add a description by clicking on the title in the top left corner
3. Add text widgets that have context on what the widgets are displaying, along with links to any helpful documentation'

See our [best practices guide](https://docs.datadoghq.com/developers/integrations/create-an-integration-dashboard/) for more info

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:13]: Please add at least one image. Usually this image is a screenshot of your dashboard or some kind of visual representation of the data being sent in

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:11]: Can you update this description to briefly describe what the integration is sending into DD?

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:15]: Please also add classifier tags for categories and submitted data types, you can find the [list here](https://docs.datadoghq.com/developers/integrations/check_references/#classifier-tags)

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:43]:
```suggestion
    "name": "EMQX",
```

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:44]: This should be the EMQX site

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:45]: We recommend having an email distribution list here and for support_email rather than an individual

**Review by anawishnoff — CHANGES_REQUESTED** [review]
First round of comments, thanks for opening this!

**Comment by zhongwencool** [comment]
@anawishnoff Could you help review this again? Also, could you check why two CI checks are reporting that they can't find python3...

File: emqx/CHANGELOG.md
Comment (emarsha94) [inline emqx/CHANGELOG.md:4]:
```suggestion
## 1.0.0 / 2023-12-20

```

File: emqx/README.md
Comment (emarsha94) [inline emqx/README.md:69]: This should be the partner support team

File: emqx/assets/logos/avatars-bot.png
Comment (emarsha94) [inline emqx/assets/logos/avatars-bot.png:1]: Please include a wordmark logo in .svg format

File: emqx/assets/dashboards/overview.json
Comment (emarsha94) [inline emqx/assets/dashboards/overview.json:1]: The description should match the overview section within the dashboard. for example: This dashboard provides an overview of your EMQX environment. It monitors node status, provides information on message status, and provides insight into various other EMQX setup observability measures.

Comment (emarsha94) [reply emqx/assets/dashboards/overview.json:1]: The title should be : EMQX - Overview

**Review by emarsha94 — CHANGES_REQUESTED** [review]

File: emqx/assets/dashboards/overview.json
Comment (zhongwencool) [reply emqx/assets/dashboards/overview.json]: @emarsha94 Thanks for reviewing. done!

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:5]:
```suggestion
[EMQX][1] is a highly scalable, open-source MQTT broker designed for IoT (Internet of Things). MQTT stands for Message Queuing Telemetry Transport, which is a lightweight, publish-subscribe network protocol that transports messages between devices.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:11]:
```suggestion
- High throughput: Capable of processing a high volume of messages efficiently.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:7]:
```suggestion
**Key features of EMQX:**
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:16]:
```suggestion
**Types of data sent to Datadog:**
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:23]:
```suggestion
**Why it's valuable:**
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:24]:
```suggestion
- Dashboard visualization: Datadog provides a visual dashboard that helps in understanding the data and metrics at a glance, which is easier than interpreting raw data.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:25]:
```suggestion
  Scalability insights: Understanding how the system performs under different loads can guide decisions on scaling the infrastructure.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:29]:
```suggestion
- Customized alerts: Setting up alerts based on specific metrics or thresholds can help in proactively managing the system.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:45]:
```suggestion
1. Edit the `emqx/conf.yaml` file, located in the `conf.d/` folder at the root of your Agent's configuration directory, to start collecting your EMQX performance data.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:14]: I recommend deleting this line or moving it under "Why it's valuable", it doesn't add meaning and seems to have been left here when other content was moved or rewritten.

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:19]:
```suggestion
- Node performance: Monitoring individual node performance in a cluster, such as latency, load, and operational metrics.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:21]:
```suggestion
- Operational health: Data about the health of the MQTT broker, including, error rates, and other critical indicators.
```

File: emqx/README.md
Comment (jhgilbert) [inline emqx/README.md:31]:
```suggestion
The integration of EMQX with Datadog enriches monitoring capabilities, providing valuable insights into the performance and health of MQTT brokers. This is especially beneficial in IoT applications where efficient, reliable, and real-time data transmission is critical.
```

**Review by jhgilbert — COMMENTED** [review]
I've made some suggestions for style -- I can loop back and approve the PR once all images are in. Thank you!

**Review by jhgilbert — APPROVED** [review]

**Comment by zhongwencool** [comment]
@jhgilbert Thanks, please help approve this, I rebase the master :0

**Comment by zhongwencool** [comment]
@jhgilbert ping

File: emqx/assets/dashboards/overview.json
Comment (emarsha94) [inline emqx/assets/dashboards/overview.json:1]: Thank you for making those changes! One last thing - for groups, can you group the similar widgets together and name the group associated to them? For example, connections. Similar to the best practices image. Thanks!

File: emqx/assets/logos/README.md
Comment (emarsha94) [inline emqx/assets/logos/README.md:1]: please remove

File: emqx/manifest.json
Comment (emarsha94) [reply emqx/manifest.json:15]: Please add one of the submitted data types from the list above

**Review by emarsha94 — CHANGES_REQUESTED** [review]

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:14]:
```suggestion
```
Don't think we need this line

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:23]: Can we move this line above so it's above "Types of data sent to Datadog"?

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:29]: This step isn't needed since all Datadog users should have the agent installed
```suggestion
```

File: emqx/README.md
Comment (anawishnoff) [inline emqx/README.md:31]:
```suggestion
Manually install the EMQX check (note that [instructions may change based on your environment][2]):
```

File: emqx/manifest.json
Comment (anawishnoff) [reply emqx/manifest.json:11]: @zhongwencool There are still a few unresolved comments on this file - please address these as well. Thank you!

**Review by anawishnoff — CHANGES_REQUESTED** [review]
A few more comments from me, apologies for the delay here! Looks like there are a few unresolved previous comments on the manifest.json file as well.

**Comment by zhongwencool** [comment]
@anawishnoff Thank you very much for helping to review the code. Now, all the suggestions should be addressed and incorporated.

File: emqx/assets/dashboards/overview.json
Comment (anawishnoff) [inline emqx/assets/dashboards/overview.json:1]: Looks like this dashboard is still missing a description

Comment (anawishnoff) [reply emqx/assets/dashboards/overview.json:1]: Also, please ensure that the title is in all capital letters for the dashboard

**Review by anawishnoff — CHANGES_REQUESTED** [review]

**Comment by anawishnoff** [comment]
Hi @zhongwencool  - looks like there are still a few unresolved comments. If you click on Files Changed and go through the files, you should be able to see all open comments. Let me know if you have any questions, thank you!

**Comment by zhongwencool** [comment]
Ping 

> Hi @zhongwencool - looks like there are still a few unresolved comments. If you click on Files Changed and go through the files, you should be able to see all open comments. Let me know if you have any questions, thank you!

Ping, I add desc and image, then rebase the master, the CI failed :(

**Comment by zhongwencool** [comment]
Ping @anawishnoff @jhgilbert

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:16]:
```suggestion
        "caption": "EMQX Broker Metrics in Datadog dashboard",
```

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:45]:
```suggestion
      },
```
This may be causing the validation issues

File: emqx/manifest.json
Comment (zhongwencool) [reply emqx/manifest.json:15]: I already add dashboard image in https://github.com/DataDog/integrations-extras/pull/2175/files#diff-3ebbdbe20cb2cc14b5c8cbf5c24d2054533015d3fee65549da90fbec7084c390 , It's not enough or ?

Comment (anawishnoff) [reply emqx/manifest.json:15]: Since you made changes to the dashboard after taking this screenshot, this image will need to be updated to show the updated dashboard

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:19]:
```suggestion
    ],
```This may fix your invalid JSON error

**Comment by zhongwencool** [comment]
ping @anawishnoff

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:11]: Please update this description to better describe the types of data being sent in, i.e. Collect performance and health data on MQTT brokers.

**Review by anawishnoff — CHANGES_REQUESTED** [review]
there are still a few open comments in the manifest.json file - once those are addressed this looks good to me. Thanks!

File: emqx/manifest.json
Comment (zhongwencool) [reply emqx/manifest.json]: done :0

**Comment by zhongwencool** [comment]
ping @anawishnoff

**Comment by anawishnoff** [comment]
Hey @zhongwencool, can you merge master into your branch? That may help us get to the bottom of why CI is failing here

**Comment by FlorentClarret** [comment]
> Hey @zhongwencool, can you merge master into your branch? That may help us get to the bottom of why CI is failing here

Hi @anawishnoff I opened https://github.com/DataDog/integrations-extras/pull/2273 to fix this, I'll merge this very soon

**Comment by anawishnoff** [comment]
@FlorentClarret Thank you! can we re-run CI?

**Comment by zhongwencool** [comment]
I run `ddev test --cov --junit emqx` command locally, the result is passed, I don't know why the CI is failed @anawishnoff

**Comment by FlorentClarret** [comment]
> I run `ddev test --cov --junit emqx` command locally, the result is passed, I don't know why the CI is failed @anawishnoff

This is a problem with pytest 8 that got released. I'll fix that today and release a new version. Nothing to do on your side, sorry for the inconvenience

**Comment by zhongwencool** [comment]
@anawishnoff All the CI builds were successful. What else do i need to do?

File: emqx/manifest.json
Comment (anawishnoff) [reply emqx/manifest.json:15]: Hi @zhongwencool, please address this feedback

File: emqx/manifest.json
Comment (anawishnoff) [inline emqx/manifest.json:11]:
```suggestion
    "description": "Collect data on the performance, health, message throughput, and message latency of MQTT brokers",
```
Feel free to revise if this is not accurate

**Review by anawishnoff — CHANGES_REQUESTED** [review]
Please see inline comments - just a few more things on my end are needed before approving.

This is not blocking, but if you have any more images of the EMQX platform or images that would be relevant to this integration , please add them as well.

**Comment by zhongwencool** [comment]
> Please see inline comments - just a few more things on my end are needed before approving.
> 
> This is not blocking, but if you have any more images of the EMQX platform or images that would be relevant to this integration , please add them as well.

I don't have these pictures at the moment. I would like to see the effect first and then add them later when I submit the PR. This PR has been taking too long.

File: emqx/hatch.toml
Comment (FlorentClarret) [inline emqx/hatch.toml:4]:
```suggestion
python = ["3.11"]
```

The agent 7.51+ is now shipped with python 3.11 so I think it would be better to bump this version

File: emqx/pyproject.toml
Comment (FlorentClarret) [inline emqx/pyproject.toml:52]:
```suggestion
```

That's not needed anymore, we don't use tox

File: emqx/setup.py
Comment (FlorentClarret) [inline emqx/setup.py:66]:
```suggestion
        'Programming Language :: Python :: 3.11',
```

File: emqx/tests/test_emqx.py
Comment (FlorentClarret) [inline emqx/tests/test_emqx.py:128]: Nice test!

File: emqx/datadog_checks/emqx/check.py
Comment (FlorentClarret) [inline emqx/datadog_checks/emqx/check.py:9]: I need to talk to Ana about this specific line. I'll let you know!

**Review by FlorentClarret — CHANGES_REQUESTED** [review]
Overall looks good to me! I left some nit comments and minor stuff, thanks.

File: emqx/datadog_checks/emqx/check.py
Comment (FlorentClarret) [reply emqx/datadog_checks/emqx/check.py:9]: All good for me!

**Review by anawishnoff — COMMENTED** [review]
There is only one [open comment](https://github.com/DataDog/integrations-extras/pull/2175/files#r1453870195) left before I can approve on my end - please address  @zhongwencool. Thank you!

File: emqx/manifest.json
Comment (zhongwencool) [reply emqx/manifest.json:15]: Is there any example of this picture to refer to? How many should be given exactly?
@anawishnoff

**Comment by zhongwencool** [comment]
> There is only one [open comment](https://github.com/DataDog/integrations-extras/pull/2175/files#r1453870195) left before I can approve on my end - please address @zhongwencool. Thank you!

Is there any example of this picture to refer to? How many should be given exactly? @anawishnoff I sincerely hope that this PR can be merged this week.

**Review by anawishnoff — APPROVED** [review]

**Comment by zhongwencool** [comment]
> Hi @zhongwencool , it looks like there are still open comments from @FlorentClarret . Please address these as well. Thanks!

Done

**Review by FlorentClarret — APPROVED** [review]

**Review by FlorentClarret — APPROVED** [review]

**Comment by zhongwencool** [comment]
@anawishnoff @FlorentClarret Thanks for reviewing and approved, I rebase the latest master, need another approval :0

**Review by emarsha94 — APPROVED** [review]

**Review by nmuesch — APPROVED** [review]
Looks good to me. I just left one inline comment about a type in the manifest's `description` field.

**Pull request merged**
